backend/danswer/connectors/teams/connector.py

```diff
@@ -0,0 +1,281 @@
+            if message.properties["from"]["user"]:
+                if message.properties["from"]["user"]["displayName"]:
+                    message_sender = message.properties["from"]["user"]["displayName"]
+                    # if its not a duplicate, add it to the list
+                    if message_sender not in [
+                        member.display_name for member in post_members_list
+                    ]:
+                        post_members_list.append(
+                            BasicExpertInfo(display_name=message_sender)
+                        )
+
+    # if there are no found post members, grab the members from the parent channel
+    if not post_members_list:
+        post_members_list = _extract_channel_members(channel)
+
+    if not thread_text:
+        return None
+
+    semantic_string = _construct_semantic_identifier(channel, top_message)
+    if not semantic_string:
+        return None
+
+    post_id = top_message.properties["id"]
+    web_url = top_message.web_url
+
+    doc = Document(
+        id=post_id,
+        sections=[Section(link=web_url, text=thread_text)],
+        source=DocumentSource.TEAMS,
+        semantic_identifier=semantic_string,
+        title="",  # teams threads don't really have a "title"
+        doc_updated_at=most_recent_message_datetime,
+        primary_owners=post_members_list,
+        metadata={},
+    )
+    return doc
+
+
+class TeamsConnector(LoadConnector, PollConnector):
+    def __init__(
+        self,
+        batch_size: int = INDEX_BATCH_SIZE,
+        teams: list[str] = [],
+    ) -> None:
+        self.batch_size = batch_size
+        self.graph_client: GraphClient | None = None
+        self.requested_team_list: list[str] = teams
+
+    def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
+        teams_client_id = credentials["teams_client_id"]
+        teams_client_secret = credentials["teams_client_secret"]
+        teams_directory_id = credentials["teams_directory_id"]
+
+        def _acquire_token_func() -> dict[str, Any]:
+            """
+            Acquire token via MSAL
+            """
+            authority_url = f"https://login.microsoftonline.com/{teams_directory_id}"
+            app = msal.ConfidentialClientApplication(
+                authority=authority_url,
+                client_id=teams_client_id,
+                client_credential=teams_client_secret,
+            )
+            token = app.acquire_token_for_client(
+                scopes=["https://graph.microsoft.com/.default"]
+            )
+            return token
+
+        self.graph_client = GraphClient(_acquire_token_func)
+        return None
+
+    def _get_all_teams(self) -> list[Team]:
+        if self.graph_client is None:
+            raise ConnectorMissingCredentialError("Teams")
+
+        teams_list: list[Team] = []
+
+        teams = self.graph_client.teams.get().execute_query()
+
+        if len(self.requested_team_list) > 0:
```

Comment:
logic can be cleaner here, like with a list comprehension, you can ask the AI to write you one here probably